### PR TITLE
Improve AI invasion handling

### PR DESCRIPTION
**Pull request by Grummel7**
This should fix the three points described in #3745. It should conquer more successful and waste less troopers now.
The AI still builds too many troopers (priority calculation sometimes produces strange results) and may ignore good planets (planetary rating also needs work), but these are for another PR.

### CODE REVIEW

File: default/python/AI/AIFleetMission.py [hunk 1]
```diff
@@ -268,7 +268,19 @@ def _check_abort_mission(self, fleet_order: AIFleetOrder):
...
+                    debug(f"Abandoning invasion for {planet}, it already belongs to us")
+                    return True
+                if planet.currentMeterValue(fo.meterType.maxShield):
+                    military_support_fleets = MilitaryAI.get_military_fleets_with_target_system(planet.systemID)
+                    if not military_support_fleets:
+                        # Maybe try again later...
+                        debug(
+                            f"Abandoning Invasion mission for {planet} because target has nonzero max shields "
+                            f"and there is no military fleet assigned to secure the target system."
+                        )
```
Comment (Cjkjvfnby) [inline default/python/AI/AIFleetMission.py:283]: Please do not make big and complicated functions bigger and more complicated.

Just extract this code to a method. `_check_abort_mission_for_invarions_fleet`.

PS.  good OOP pattern is to replace these is instance calls with a single call of the fleet_order.should_abort() and move code from branches to classes.  But this is out of scope of this PR.

File: default/python/AI/AIFleetMission.py [hunk 2]
```diff
@@ -500,15 +512,15 @@ def issue_fleet_orders(self):
                         return  # colonize order must not have completed yet
                 clear_all = True
                 last_sys_target = INVALID_ID
+                secure_targets = set(
+                    AIstate.colonyTargetedSystemIDs
+                    + AIstate.outpostTargetedSystemIDs
+                    + AIstate.invasionTargetedSystemIDs
+                )
```
Comment (Cjkjvfnby) [inline default/python/AI/AIFleetMission.py:519]: I don't have old python installed.  This works in Pyhton3.9, but I am not sure if it works in 3.6

You can create set with unpacking variable.
```python
a = [1,2, 3]
b = [3,4, 5]
{*a, *b}
{1, 2, 3, 4, 5}
```

File: default/python/AI/AIFleetMission.py [hunk 3]
```diff
@@ -533,6 +545,11 @@ def issue_fleet_orders(self):
                     last_sys_target = self.target.id
                     debug("Check if PROTECT_REGION mission with target %d is finished.", last_sys_target)
 
+                # for SECURE missions, only release fleet if no settle or invade mission is ongoing
+                if self.type == MissionType.SECURE and self.target.id in secure_targets:
+                    debug(f"Secure mission for {self.target} not yet finished")
```
Comment (Cjkjvfnby) [inline default/python/AI/AIFleetMission.py:550]: Maybe print other info about the mission? For example fleet.   You can try to print `self`.  This will cal `__str__` method implicitly.

File: default/python/AI/InvasionAI.py [hunk 4]
```diff
@@ -543,27 +527,45 @@ def evaluate_invasion_planet(planet_id):
...
+        f" - popval: {colony_base_value:.1f}\n"
+        f" - bldval: {bld_tally}\n"
+        f" - enemyval: {enemy_val}\n"
+        f" - system secured: {system_secured}"
     )
-    debug(" - system secured: %s" % system_secured)
     return [planet_score, planned_troops]
 
 
+def secure_system(system_id: SystemId, secure_now: bool):
```
Comment (Cjkjvfnby) [inline default/python/AI/InvasionAI.py]: You could specify return type with annotations.
```suggestion
def secure_system(system_id: SystemId, secure_now: bool) -> bool:
```

File: default/python/AI/InvasionAI.py [hunk 5]
```diff
@@ -543,27 +527,45 @@ def evaluate_invasion_planet(planet_id):
...
+    """Check whether we have any military fleets in the system.
+    If secure_now is True, set one of the fleets to MissionType.SECURE
+    If secure_now is False, return true if there is any fleet that secure it.
```
Comment (Cjkjvfnby) [inline default/python/AI/InvasionAI.py]:
```suggestion
    If secure_now is False, return True if there is any fleet that secures it.
```

File: default/python/AI/InvasionAI.py [hunk 6]
```diff
@@ -543,27 +527,45 @@ def evaluate_invasion_planet(planet_id):
...
+        fleet = universe.getFleet(mission.fleet.id)
+        if not fleet or fleet.systemID != system_id:
+            continue
+        target_obj = mission.target.get_object()
+        if target_obj is not None and target_obj.id in secure_targets:
+            if secure_now and mission.type != MissionType.SECURE:
+                mission.set_target(MissionType.SECURE, mission.target)
+            secured = system_status.get("myFleetRating", 0) > 0
+            break
+    return secured
```
Comment (Cjkjvfnby) [inline default/python/AI/InvasionAI.py:568]: We could try to remove the `secured` variable. 

```suggestion
            return secured
    return False
```

Should we stop on the first fleet to we should keep trying?
If this felt is not

**Review by Cjkjvfnby — COMMENTED** [review]
Looks great, a couple of minor comments.

File: default/python/AI/AIFleetMission.py (hunk 2 above)
Comment (Grummel7) [reply default/python/AI/AIFleetMission.py:519]: Well, I just moved this code, because I needed the set outside if clause. There are surely more important things to refactor than this.

File: default/python/AI/AIFleetMission.py (hunk 3 above)
Comment (Grummel7) [reply default/python/AI/AIFleetMission.py:550]: There is lots of debug outputs about the mission before, reading the logs you do not miss that information.

File: default/python/AI/InvasionAI.py (hunk 5 above)
Comment (Grummel7) [reply default/python/AI/InvasionAI.py]: actually it should be ... that could secure it.

File: default/python/AI/InvasionAI.py (hunk 6 above)
Comment (Grummel7) [reply default/python/AI/InvasionAI.py:568]: I know that could isn't perfect. I copied the "myFleetRating" stuff. It seems to work in by and large...
I think I should rather check whether the fleet I just set to secure actually does the job, but I have to figure out how to do that. I'll add a TBD to make that clearer.

File: default/python/AI/AIFleetMission.py (hunk 1 above)
Comment (Grummel7) [reply default/python/AI/AIFleetMission.py:283]: Looking a fleet orders I wonder whether we could simply use fleet_order.is_valid. There is definitely some code duplication here.
But that is not a trivial refactoring, I suggest merging this as it is. _check_abort_mission is not that bad yet.

**Review by Cjkjvfnby — APPROVED** [review]

**Comment by geoffthemedio** [comment]
@Cjkjvfnby why did you approve but not merge?

**Comment by Cjkjvfnby** [comment]
> @Cjkjvfnby why did you approve but not merge?

Sorry.  At my job author of PR merges it, so I constantly forget to merge in the FreeOrion project.
